File: _/aws-ch2-samples/src/com/packtpub/a3ws/ch2/samples/association/RemoveChild.java
package com.packtpub.a3ws.ch2.samples.association;

import java.rmi.RemoteException;

import org.alfresco.webservice.repository.RepositoryFault;
import org.alfresco.webservice.repository.UpdateResult;
import org.alfresco.webservice.types.CML;
import org.alfresco.webservice.types.CMLAddChild;
import org.alfresco.webservice.types.CMLCreate;
import org.alfresco.webservice.types.CMLRemoveChild;
import org.alfresco.webservice.types.NamedValue;
import org.alfresco.webservice.types.ParentReference;
import org.alfresco.webservice.types.Predicate;
import org.alfresco.webservice.types.Reference;
import org.alfresco.webservice.types.Store;
import org.alfresco.webservice.util.AuthenticationUtils;
import org.alfresco.webservice.util.Constants;
import org.alfresco.webservice.util.Utils;
import org.alfresco.webservice.util.WebServiceFactory;

public class RemoveChild {

	/**
	 * @param args
	 * @throws RemoteException 
	 * @throws RepositoryFault 
	 */
	public static void main(String[] args) throws RepositoryFault, RemoteException {
		AuthenticationUtils.startSession("admin", "admin");
		Store spacesStore = new Store(Constants.WORKSPACE_STORE, "SpacesStore");
		String name = "AWS Book - Remove Child "+System.currentTimeMillis();
		String spaceNameForMove = "AWS Book - Remove Child Space Sample "+System.currentTimeMillis();
		String description = "This is a content created with a sample of the book";
		
		try {

			//parent for the new node
			ParentReference parentForNode = new ParentReference(
					spacesStore, 
					null,
					"/app:company_home",
					Constants.ASSOC_CONTAINS,
					"{" + Constants.NAMESPACE_CONTENT_MODEL + "}"+name);
			
			//parent for the new space
			ParentReference parentForSpace = new ParentReference(
					spacesStore, 
					null,
					"/app:company_home",
					Constants.ASSOC_CONTAINS,
					"{" + Constants.NAMESPACE_CONTENT_MODEL + "}"+spaceNameForMove);
			
			
			//build properties
			NamedValue[] properties = buildCustomProperties(name,name,description);
			NamedValue[] propertiesForSpace = buildCustomProperties(spaceNameForMove,spaceNameForMove,description);
			
			//create a node
			CMLCreate create = new CMLCreate();
			create.setId("1");
			create.setParent(parentForNode);
			create.setType(Constants.TYPE_CONTENT);
			create.setProperty(properties);
			
			//create a space
			CMLCreate createSpace = new CMLCreate();
			createSpace.setId("2");
			createSpace.setParent(parentForSpace);
			createSpace.setType(Constants.TYPE_FOLDER);
			createSpace.setProperty(propertiesForSpace);
			
			//build the CML object
			CML cmlAdd = new CML();
			cmlAdd.setCreate(new CMLCreate[]{create, createSpace});
	        
	        //perform a CML update to create nodes
	        UpdateResult[] result = WebServiceFactory.getRepositoryService().update(cmlAdd);
	        

	        System.out.println("Node created: "+result[0].getDestination().getPath());
	        System.out.println("Space created: "+result[1].getDestination().getPath());
	        
	        //create a predicate with the first CMLCreate result
	        Reference referenceForNode = result[0].getDestination();
	        Predicate sourcePredicate = new Predicate(new Reference[]{referenceForNode}, spacesStore, null);
	        
	        //create a reference from the second CMLCreate performed for space
	        Reference referenceForTargetSpace = result[1].getDestination();
	        
	        //reference for the target space
	        ParentReference targetSpace = new ParentReference();
	        targetSpace.setStore(spacesStore);
	        targetSpace.setPath(referenceForTargetSpace.getPath());
	        targetSpace.setAssociationType(Constants.ASSOC_CONTAINS);
	        targetSpace.setChildName(name);
	        
	        //add child
	        CMLAddChild addChild = new CMLAddChild();
	        addChild.setWhere(sourcePredicate);
	        addChild.setTo(targetSpace);
	        
	        CML cmlAddChild = new CML();
	        cmlAddChild.setAddChild(new CMLAddChild[]{addChild});
	        
	        //perform a CML update to add a child node
	        UpdateResult[] resultAddChild = WebServiceFactory.getRepositoryService().update(cmlAddChild);
	        
	        System.out.println("Node: "+referenceForNode.getPath()+" added to space: "+targetSpace.getPath());
	        
	        Reference refUpdate = resultAddChild[0].getDestination();
	        Predicate nodeToRemove = new Predicate(new Reference[]{refUpdate}, spacesStore, null);
	        
	        //remove child
	        CMLRemoveChild removeChild = new CMLRemoveChild();
	        removeChild.setFrom(targetSpace);
	        removeChild.setWhere(nodeToRemove);
	        
	        CML cmlRemoveChild = new CML();
	        cmlRemoveChild.setRemoveChild(new CMLRemoveChild[]{removeChild});
	        
	        //perform a CML update to remove the node
	        WebServiceFactory.getRepositoryService().update(cmlRemoveChild);
	        
	        System.out.println("Node: "+nodeToRemove.getNodes(0).getPath()+" removed from space: "+referenceForTargetSpace.getPath());

		} finally {
			AuthenticationUtils.endSession();
		}

	}

	private static NamedValue[] buildCustomProperties(String name, String title, String description) {
		NamedValue[] properties = new NamedValue[3];
		properties[0] = Utils.createNamedValue(Constants.PROP_NAME, name);
		properties[1] = Utils.createNamedValue(Constants.PROP_TITLE, title);
		properties[2] = Utils.createNamedValue(Constants.PROP_DESCRIPTION, description);
		return properties;
	}

}

File: _/1520_05_PreFinals_samples/src/Samples/Association/RemoveChild.cs
using System;
using Alfresco;
using Alfresco.RepositoryWebService;
using NUnit.Framework;
using Samples.Vo;

namespace Samples.Association
{
	[TestFixture]
	public class RemoveChild
	{
		[TestCase]
		public void CanRemoveChildNodes()
		{
			AuthenticationUtils.startSession("admin", "admin");
			Store spacesStore = new Store(StoreEnum.workspace, "SpacesStore");
			String name = "AWS Book - Remove Child " + DateTime.Now.Ticks;
			String spaceNameForMove = "AWS Book - Remove Child Space Sample " + DateTime.Now.Ticks;
			String description = "This is a content created with a sample of the book";
			
			//custom value object
			CreateSampleVO createSampleVo = Builder.BuildCreateSampleVO(name, name, description);
			CreateSampleVO createFolderSampleVo = Builder.BuildCreateSampleVO(spaceNameForMove, spaceNameForMove, description);
			
			try {
	
				//parent for the new node
				ParentReference parentForNode = new ParentReference(
						spacesStore, 
						null,
						"/app:company_home",
						Constants.ASSOC_CONTAINS,
						"{" + Constants.NAMESPACE_CONTENT_MODEL + "}" + name
				);
				
				//parent for the new space
				ParentReference parentForSpace = new ParentReference(
						spacesStore, 
						null,
						"/app:company_home",
						Constants.ASSOC_CONTAINS,
						"{" + Constants.NAMESPACE_CONTENT_MODEL + "}" + spaceNameForMove
				);
				
				
				//build properties
				NamedValue[] properties = Builder.BuildCustomProperties(createSampleVo);
				NamedValue[] propertiesForSpace = Builder.BuildCustomPropertiesForSpace(createFolderSampleVo);
				
				//create a node
				CMLCreate create = new CMLCreate();
				create.id = "1";
				create.parent = parentForNode;
				create.type = Constants.TYPE_CONTENT;
				create.property = properties;
				
				//create a space
				CMLCreate createSpace = new CMLCreate();
				createSpace.id = "2";
				createSpace.parent = parentForSpace;
				createSpace.type = Constants.TYPE_FOLDER;
				createSpace.property = propertiesForSpace;
				
				//build the CML object
				CML cmlAdd = new CML();
				cmlAdd.create = new CMLCreate[]{ create, createSpace };
		        
		        //perform a CML update to create nodes
		        UpdateResult[] result = WebServiceFactory.getRepositoryService().update(cmlAdd);
		        
				String expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Remove_x0020_Child_x0020_";
				Assert.IsTrue(result[0].destination.path.StartsWith(expectedPath));
		        
		        expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Remove_x0020_Child_x0020_Space_x0020_Sample_x0020_";
		        Assert.IsTrue(result[1].destination.path.StartsWith(expectedPath));
		        
		        //create a predicate with the first CMLCreate result
		        Reference referenceForNode = result[0].destination;
		        Predicate sourcePredicate = new Predicate(new Reference[]{ referenceForNode }, spacesStore, null);
		        
		        //create a reference from the second CMLCreate performed for space
		        Reference referenceForTargetSpace = result[1].destination;
		        
		        //reference for the target space
		        ParentReference targetSpace = new ParentReference();
		        targetSpace.store = spacesStore;
		        targetSpace.path = referenceForTargetSpace.path;
		        targetSpace.associationType = Constants.ASSOC_CONTAINS;
		        targetSpace.childName = name;
		        
		        //add child
		        CMLAddChild addChild = new CMLAddChild();
		        addChild.where = sourcePredicate;
		        addChild.to = targetSpace;
		        
		        CML cmlAddChild = new CML();
		        cmlAddChild.addChild = new CMLAddChild[]{ addChild };
		        
		        //perform a CML update to add a child node
		        UpdateResult[] resultAddChild = WebServiceFactory.getRepositoryService().update(cmlAddChild);
		        
		        expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Remove_x0020_Child_x0020_";
		        Assert.IsTrue(referenceForNode.path.StartsWith(expectedPath));
		        
		        expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Remove_x0020_Child_x0020_Space_x0020_Sample_x0020_";
		        Assert.IsTrue(targetSpace.path.StartsWith(expectedPath));
		        
		        Reference refUpdate = resultAddChild[0].destination;
		        Predicate nodeToRemove = new Predicate(new Reference[]{ refUpdate }, spacesStore, null);
		        
		        //remove child
		        CMLRemoveChild removeChild = new CMLRemoveChild();
		        removeChild.Item = targetSpace;
		        removeChild.where = nodeToRemove;
		        
		        CML cmlRemoveChild = new CML();
		        cmlRemoveChild.removeChild = new CMLRemoveChild[]{ removeChild };
		        
		        //perform a CML update to remove the node
		        WebServiceFactory.getRepositoryService().update(cmlRemoveChild);
		        
		        expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Remove_x0020_Child_x0020_";
		        Reference firtsItem = nodeToRemove.Items[0] as Reference;
		        Assert.IsTrue( firtsItem.path.StartsWith(expectedPath));
		        
		        expectedPath = "/app:company_home/cm:AWS_x0020_Book_x0020_-_x0020_Remove_x0020_Child_x0020_Space_x0020_Sample_x0020_";
		        Assert.IsTrue(referenceForTargetSpace.path.StartsWith(expectedPath));
	
			} finally {
				AuthenticationUtils.endSession();
			}
		}
	}
}
